hal/modem_api.h

```diff
@@ -0,0 +1,88 @@
+/* mbed Microcontroller Library
```

Comment:
The vast bulk of the implementation is very generic, so for future extension, to incorporate off-board devices with generic command sets, we could split this class's implementation with a base such that an off-board driver can inherit from that base and provide alternatives to the HAL calls and the UART FileHandle.

That seems reasonable, in the absence of a standard mechanism to treat on-board and off-board devices equivalently. 

But I would still retain the base on-board HAL implementation - just as applications don't need to specify "STMSPI", "K64FSPI", "uBloxSPI", I don't see why applications should be forced to use "UbloxCellularInterface" or "XXXCellularInterface", as long as they're on-board.

It's annoying that applications have to specify hardware when off-board, but I don't think we should be dragging on-board devices down to that level.

So I think such a split could happen after this PR - the split wouldn't change the HAL or application interfaces for on-board modems, it would just be providing alternative attachment hooks for off-board modems.